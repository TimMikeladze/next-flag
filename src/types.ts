/* global RequestInit */

import { ReactNode } from 'react';

export interface Condition {
  enabled: boolean;
  name: string;
}

export interface Environment {
  enabled: boolean;
  name: string;
}

export interface Feature {
  conditions: Record<string, Condition>;
  enabled: boolean;
  environments: Record<string, Environment>;
  name: string;
}

export interface Features {
  [slug: string]: Feature;
}

export type Context = Record<string, unknown>;

export interface NextFlagOptionsPath {
  conditions?: Record<string, (context: Context) => Promise<boolean> | boolean>;
  issue: number;
  project?: string;

  repository: string;
}

export type RevalidateTag = (key: string) => void;

export type UnstableCache = <T extends (...args: any[]) => Promise<any>>(
  cb: T,
  keyParts?: string[],
  options?: {
    revalidate?: number | false;
    tags?: string[];
  }
) => T;

export interface NextFlagOptions {
  cache?: {
    revalidateTag: RevalidateTag;
    unstable_cache: UnstableCache;
  };

  getEnvironment?: () => Promise<string> | string;

  github?: {
    secret: string;
    token?: string;
  };

  paths: NextFlagOptionsPath[];

  requestToContext?: (req: Request) => Promise<Record<string, unknown>>;

  standalone?: boolean;
}

export type GetFeatures = string[];

export interface NextFlagWebhookBody {
  action: 'created' | 'edited';
  issue: {
    body: string;
    labels: {
      color: string;
      default: boolean;
      description: string;
      id: number;
      name: string;
      node_id: string;
      url: string;
    }[];
    number: number;
    state: 'open' | 'closed';
    title: string;
    user: {
      id: number;
      login: string;
      type: 'User';
    };
  };
  repository: {
    full_name: string;
    name: string;
  };
}

export interface UseNextFlagHookProps {
  endpoint?: string;
  environment?: string;

  project?: string;

  requestInit?: RequestInit;
}

export interface NextFlagProviderProps {
  children?: ReactNode;

  endpoint?: string;
  environment?: string;
  project?: string;

  requestInit?: RequestInit;
}

export interface IsFeatureEnabledOptions {
  endpoint?: string;
  environment?: string;

  project?: string;
  requestInit?: RequestInit;
}

export interface GetFeaturesArgs {
  endpoint?: string;
  environment?: string;

  project?: string;

  requestInit?: RequestInit;
}
