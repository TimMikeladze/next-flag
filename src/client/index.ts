export * from './getFeatures';
export * from './isEnabled';
export * from './isFeatureEnabled';
export * from './getDefaultEnvironment';
