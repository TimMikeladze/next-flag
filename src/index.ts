export * from './webhook';
export * from './parser';
export * from './types';
export * from './NextFlag';
