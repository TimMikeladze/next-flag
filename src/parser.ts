/* eslint-disable no-loop-func */
/* eslint-disable no-continue */
import { parse } from '@textlint/markdown-to-ast';
import { spaceSlug } from 'space-slug';
import { Feature, Features } from './types';

interface Node {
  children?: Node[];
  depth?: number;
  raw: string;

  type: string;
}

export const parseMarkdown = (markdown: string): Features => {
  const ast = parse(markdown);

  const isFeatureBlock = (node: Node) =>
    node?.type === 'Header' && node?.depth === 2;

  const isEnabled = (node: Node) =>
    (node?.type === 'List' || node?.type === 'ListItem') &&
    node?.raw.toLowerCase().trim().startsWith('- [x]');

  const isConditionsBlock = (node: Node) =>
    node?.type === 'Header' &&
    node?.raw.toLowerCase().trim() === '#### conditions' &&
    node?.depth === 4;

  const isEnvironmentsBlock = (node: Node) =>
    node?.type === 'Header' &&
    node?.raw.toLowerCase().startsWith('### ') &&
    node?.depth === 3;

  const isList = (node: Node) => node?.type === 'List';

  const cleanStartingHash = (text: string) => text.replace(/^#+\s*/, '');

  const cleanStartingCheckbox = (text: string) =>
    text.replace(/- \[x\]\s*/, '').replace(/- \[\s+\]\s*/, '');

  const features: Features = {};

  let currentFeature: Feature | null = null;

  for (
    let featureIndex = 0;
    featureIndex < ast.children.length;
    featureIndex++
  ) {
    if (!isFeatureBlock(ast.children[featureIndex])) {
      if (isConditionsBlock(ast.children[featureIndex])) {
        let conditionsBlockIndex = featureIndex + 1;
        while (isList(ast.children[conditionsBlockIndex])) {
          for (
            let conditionIndex = 0;
            conditionIndex <
            (ast.children[conditionsBlockIndex] as Node).children!.length;
            conditionIndex++
          ) {
            const conditionNode = (ast.children[conditionsBlockIndex] as Node)
              .children?.[conditionIndex];
            if (!conditionNode) {
              continue;
            }
            const conditionName = cleanStartingCheckbox(conditionNode.raw);
            const conditionSlug = spaceSlug([conditionName]);
            const conditionEnabled = isEnabled(conditionNode);
            if (currentFeature) {
              currentFeature.conditions[conditionSlug] = {
                name: conditionName,
                enabled: conditionEnabled,
              };
            }
          }
          conditionsBlockIndex++;
        }
      }

      if (isEnvironmentsBlock(ast.children[featureIndex])) {
        const environmentName = cleanStartingHash(
          ast.children[featureIndex].raw
        );
        const environmentSlug = spaceSlug([environmentName]);
        if (currentFeature) {
          currentFeature.environments[environmentSlug] = {
            name: environmentName,
            enabled: false,
          };
        }
        let environmentsBlockIndex = featureIndex + 1;
        while (isList(ast.children[environmentsBlockIndex])) {
          for (
            let environmentIndex = 0;
            environmentIndex <
            (ast.children[environmentsBlockIndex] as Node).children!.length;
            environmentIndex++
          ) {
            const environmentNode = (
              ast.children[environmentsBlockIndex] as Node
            ).children?.[environmentIndex];
            if (!environmentNode) {
              continue;
            }
            const environmentEnabled = isEnabled(environmentNode);
            if (currentFeature) {
              currentFeature.environments[environmentSlug] = {
                name: environmentName,
                enabled: environmentEnabled,
              };
            }
          }
          environmentsBlockIndex++;
        }
      }

      continue;
    }
    const featureName = cleanStartingHash(ast.children[featureIndex].raw);
    const featureSlug = spaceSlug([featureName]);
    features[featureSlug] = {
      name: featureName,
      environments: {},
      conditions: {},
      enabled: false,
    };
    currentFeature = features[featureSlug];
    let featureBlockIndex = featureIndex + 1;
    while (isList(ast.children[featureBlockIndex] as Node)) {
      const node = ast.children[featureBlockIndex] as Node;
      const enabled = isEnabled(node);
      if (currentFeature) {
        currentFeature.enabled = enabled;
      }
      featureBlockIndex++;
    }
  }

  return features;
};
