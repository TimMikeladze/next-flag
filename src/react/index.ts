export * from './useNextFlag';
export * from './NextFlagProvider';
