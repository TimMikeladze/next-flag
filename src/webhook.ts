import crypto from 'crypto';
import { NextFlagWebhookBody } from './types';

export const verifyWebhookSignature = async (
  req: Request,
  secret: string
): Promise<NextFlagWebhookBody | false> => {
  const sigHeaderName = 'x-hub-signature-256';
  const sigHashAlg = 'sha256';
  const body = await req.text();
  const sig = Buffer.from(req.headers.get(sigHeaderName) || '', 'utf8');
  const hmac = crypto.createHmac(sigHashAlg, secret);
  const digest = Buffer.from(
    `${sigHashAlg}=${hmac.update(body).digest('hex')}`,
    'utf8'
  );

  if (sig.length !== digest.length || !crypto.timingSafeEqual(digest, sig)) {
    return false;
  }

  return JSON.parse(body);
};
