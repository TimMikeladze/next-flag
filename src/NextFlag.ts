// eslint-disable-next-line camelcase
import { NextRequest, NextResponse } from 'next/server';
import { Octokit } from '@octokit/rest';
import {
  Context,
  GetFeatures,
  NextFlagOptions,
  NextFlagOptionsPath,
  RevalidateTag,
  UnstableCache,
} from './types';
import { verifyWebhookSignature } from './webhook';
import { parseMarkdown } from './parser';
import { isEnabled } from './client/isEnabled';
import { getDefaultEnvironment } from './client';

export class NextFlag {
  private secret: string;

  private token: string;

  private paths: NextFlagOptionsPath[];

  private octokit: Octokit;

  private revalidateTag: RevalidateTag | undefined;

  private unstable_cache: UnstableCache | undefined;

  private getEnvironment: () => Promise<string> | string;

  private standalone: boolean = false;

  private requestToContext: (req: NextRequest) => Promise<Context> | Context;

  private logging: boolean = false;

  constructor(options: NextFlagOptions) {
    this.standalone = options.standalone || false;
    this.secret =
      options.github?.secret ||
      (process.env.NEXT_FLAG_WEBHOOK_SECRET as string);
    this.token =
      options.github?.token || (process.env.NEXT_FLAG_GITHUB_TOKEN as string);

    if (!this.secret) {
      console.warn(
        'NEXT_FLAG_WEBHOOK_SECRET not provided. It is suggested to configure a GitHub Webhook in order to activate NextJS caching. Without caching, the GitHub API will be queried for every feature flag check.'
      );
    }

    this.getEnvironment = options.getEnvironment || getDefaultEnvironment;

    this.requestToContext = options.requestToContext || (async () => ({}));

    this.logging = options.logging || false;

    this.octokit = new Octokit({
      auth: this.token,
    });

    this.unstable_cache = options.cache?.unstable_cache;
    this.revalidateTag = options.cache?.revalidateTag;

    this.paths = options.paths.map((x) => {
      if (!NextFlag.isValidRepo(x.repository)) {
        throw new Error('Invalid repo');
      }
      return {
        project: NextFlag.pathToProject(x),
        repository: x.repository.toLowerCase(),
        issue: x.issue,
        conditions: x.conditions || {},
      };
    });

    if (!this.paths.length) {
      throw new Error('At least one path must be provided');
    }

    if (this.paths.length > 1) {
      const set = new Set(this.paths.map((path) => path.project));
      if (set.size !== this.paths.length) {
        throw new Error('Each path must have a unique project');
      }
    }
  }

  private static pathToProject(path: NextFlagOptionsPath) {
    if (path.project) {
      return path.project;
    }
    return `${path.repository.toLowerCase()}/${path.issue}`;
  }

  private static isValidRepo(repo: string) {
    return repo.split('/').length === 2;
  }

  private static getRepoOwner(repo: string) {
    return repo.split('/')[0];
  }

  private static getRepoName(repo: string) {
    return repo.split('/')[1];
  }

  private async getCachedFeatures(project: string, environment: string) {
    const fn = async (pathProject: string) => {
      const foundPath = this.paths.find((path) => path.project === pathProject);
      if (!foundPath) {
        throw new Error('Path not found');
      }
      const owner = NextFlag.getRepoOwner(foundPath.repository);
      const repo = NextFlag.getRepoName(foundPath.repository);
      const { data } = await this.octokit.rest.issues.get({
        owner,
        repo,
        issue_number: foundPath.issue,
      });
      if (!data.body) {
        throw new Error('Issue body not found');
      }
      const features = parseMarkdown(data.body);

      if (this.logging) {
        console.log(
          `Fetching from GitHub features for ${project} in ${environment}`
        );
        console.log(JSON.stringify(features, null, 2));
      }

      return { ...features };
    };

    // No webhook enabled, therefore no caching.
    if (!this.secret) {
      return fn(project);
    }

    // eslint-disable-next-line camelcase
    const unstable_cache = (this.unstable_cache ||
      (await import('next/cache').then(
        (x) => x.unstable_cache
      ))) as UnstableCache;

    return unstable_cache(fn, [project, environment], {
      tags: [project, environment],
    })(project);
  }

  public async GET(req: NextRequest) {
    const project = req.nextUrl.searchParams.get('project') as string;
    let environment = req.nextUrl.searchParams.get('environment') as string;

    // If not running in standalone mode, the environment must be sourced from the server environment.
    if (environment && !this.standalone) {
      environment = getDefaultEnvironment();
    }

    let context;
    if (this.requestToContext) {
      context = await this.requestToContext(req);
    }

    return NextResponse.json(
      await this.getFeatures({
        environment,
        project,
        context,
      })
    );
  }

  public async isFeatureEnabled(
    feature: string | string[],
    options: {
      context?: Context;
      environment?: string;
      project?: string;
    } = {}
  ): Promise<boolean> {
    const features = await this.getFeatures({
      environment: options.environment,
      project: options.project,
      context: options.context || {},
    });
    return isEnabled(feature, features);
  }

  public async getFeatures(
    options: {
      context?: Context;
      environment?: string;
      project?: string;
    } = {}
  ): Promise<GetFeatures> {
    const { environment, project } = options;

    const requestedProject = (
      project || (process.env.NEXT_PUBLIC_NEXT_FLAG_PROJECT as string)
    ).toLowerCase();

    const requestedEnvironment = (
      environment || (await this.getEnvironment())
    ).toLowerCase();

    if (this.paths.length > 1 && !requestedProject) {
      throw new Error(
        'project must be provided when multiple paths are defined'
      );
    }
    if (
      requestedProject &&
      !this.paths.find((path) => path.project === requestedProject)
    ) {
      throw new Error('project not found');
    }
    const path =
      this.paths.find((x) => x.project === requestedProject) || this.paths[0];

    if (!path.project) {
      throw new Error('path not found');
    }

    const cachedFeatures = await this.getCachedFeatures(
      path.project,
      requestedEnvironment
    );

    const features = Object.keys(cachedFeatures).reduce((acc, curr) => {
      const feature = cachedFeatures[curr].enabled;
      if (!feature) {
        return acc;
      }
      const hasEnvironments =
        Object.keys(cachedFeatures[curr].environments).length > 0;
      if (!feature && !hasEnvironments) {
        return acc;
      }
      if (feature && !hasEnvironments) {
        acc.add(curr);
        return acc;
      }
      const foundEnvironment =
        cachedFeatures[curr].environments[requestedEnvironment];
      if (!foundEnvironment) {
        return acc;
      }
      if (foundEnvironment.enabled) {
        acc.add(curr);
        return acc;
      }
      return acc;
    }, new Set());

    const enabledFeatures = Array.from(features) as GetFeatures;

    const enabledFeaturesWithConditions = [] as GetFeatures;

    await Promise.all(
      enabledFeatures.map(async (feature) => {
        const conditionsForPath = path.conditions || {};
        const conditionsForFeature = cachedFeatures[feature].conditions;
        const conditionKeys = Object.keys(conditionsForFeature);
        if (conditionKeys.length === 0) {
          enabledFeaturesWithConditions.push(feature);
          return;
        }

        const computedConditions = await Promise.all(
          conditionKeys.map(async (conditionKey) => {
            const condition = conditionsForFeature[conditionKey];
            if (!condition.enabled) {
              enabledFeaturesWithConditions.push(feature);
              return true;
            }
            const conditionFn = conditionsForPath[conditionKey];
            const result =
              (await conditionFn?.(options.context || {})) || false;
            return result;
          })
        );

        if (computedConditions.every((x) => x === true)) {
          enabledFeaturesWithConditions.push(feature);
        }
      })
    );

    return enabledFeaturesWithConditions;
  }

  public async POST(req: NextRequest) {
    return this.githubWebhook(req);
  }

  private async githubWebhook(req: NextRequest) {
    if (!this.secret) {
      return NextResponse.json(
        { error: 'Secret not found. GitHub webhook disabled.' },
        { status: 200 }
      );
    }

    const body = await verifyWebhookSignature(req, this.secret);
    if (!body) {
      return NextResponse.json({ error: 'Invalid signature' }, { status: 401 });
    }

    const foundPath = this.paths.find(
      (path) =>
        path.repository.toLowerCase() ===
          body.repository.full_name.toLowerCase() &&
        path.issue === body.issue.number
    );

    if (!foundPath) {
      return NextResponse.json({ error: 'Path not found' }, { status: 200 });
    }

    if (!foundPath.project) {
      throw new Error('Path project not found');
    }

    const revalidateTag = (this.revalidateTag ||
      (await import('next/cache').then(
        (x) => x.revalidateTag
      ))) as RevalidateTag;

    revalidateTag(foundPath.project);

    if (this.logging) {
      console.log('Revalidated cache for', foundPath.project);

      const markdown = body.issue.body;
      const features = parseMarkdown(markdown);

      console.log(JSON.stringify(features, null, 2));
    }

    return NextResponse.json({ success: true });
  }
}
