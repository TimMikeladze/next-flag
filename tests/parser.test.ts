import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { resolve } from 'path';
import { parseMarkdown } from '../src/parser';

const loadMarkdown = (filename: string) =>
  readFileSync(resolve(`tests/markdown/${filename}`), 'utf-8');

describe('parseMarkdown', () => {
  it('full-example.md', () => {
    const markdown = loadMarkdown('full-example.md');

    expect(parseMarkdown(markdown)).toMatchSnapshot();
  });
  it('only-features.md', () => {
    const markdown = loadMarkdown('only-features.md');

    expect(parseMarkdown(markdown)).toMatchSnapshot();
  });
  it('no-features-enabled.md', () => {
    const markdown = loadMarkdown('no-features-enabled.md');

    expect(parseMarkdown(markdown)).toMatchSnapshot();
  });
  it('several-features-multi-env.md', () => {
    const markdown = loadMarkdown('several-features-multi-env.md');

    expect(parseMarkdown(markdown)).toMatchSnapshot();
  });
});
