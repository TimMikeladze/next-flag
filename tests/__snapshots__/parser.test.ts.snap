// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseMarkdown > full-example.md 1`] = `
{
  "show-footer": {
    "conditions": {},
    "enabled": true,
    "environments": {},
    "name": "Show footer",
  },
  "show-install-script": {
    "conditions": {
      "condition-5": {
        "enabled": false,
        "name": "Condition 5",
      },
      "condition-6": {
        "enabled": true,
        "name": "Condition 6",
      },
    },
    "enabled": true,
    "environments": {
      "production": {
        "enabled": true,
        "name": "Production",
      },
    },
    "name": "Show install script",
  },
  "show-logo": {
    "conditions": {
      "condition-1": {
        "enabled": false,
        "name": "Condition 1",
      },
      "condition-2": {
        "enabled": true,
        "name": "Condition 2",
      },
    },
    "enabled": true,
    "environments": {
      "development": {
        "enabled": false,
        "name": "Development",
      },
      "preview": {
        "enabled": true,
        "name": "Preview",
      },
      "production": {
        "enabled": false,
        "name": "Production",
      },
    },
    "name": "Show logo",
  },
}
`;

exports[`parseMarkdown > no-features-enabled.md 1`] = `
{
  "show-footer": {
    "conditions": {},
    "enabled": false,
    "environments": {},
    "name": "Show footer",
  },
  "show-install-script": {
    "conditions": {},
    "enabled": false,
    "environments": {},
    "name": "Show install script",
  },
  "show-logo": {
    "conditions": {},
    "enabled": false,
    "environments": {},
    "name": "Show logo",
  },
}
`;

exports[`parseMarkdown > only-features.md 1`] = `
{
  "show-footer": {
    "conditions": {},
    "enabled": true,
    "environments": {},
    "name": "Show footer",
  },
  "show-install-script": {
    "conditions": {},
    "enabled": false,
    "environments": {},
    "name": "Show install script",
  },
  "show-logo": {
    "conditions": {},
    "enabled": true,
    "environments": {},
    "name": "Show logo",
  },
}
`;

exports[`parseMarkdown > several-features-multi-env.md 1`] = `
{
  "show-footer": {
    "conditions": {},
    "enabled": true,
    "environments": {
      "production": {
        "enabled": true,
        "name": "Production",
      },
    },
    "name": "Show footer",
  },
  "show-install-script": {
    "conditions": {},
    "enabled": false,
    "environments": {
      "development": {
        "enabled": true,
        "name": "Development",
      },
      "preview": {
        "enabled": false,
        "name": "Preview",
      },
    },
    "name": "Show install script",
  },
  "show-logo": {
    "conditions": {},
    "enabled": true,
    "environments": {},
    "name": "Show logo",
  },
}
`;
